pkg/clients/ec2/vpccidrblock.go

```diff
@@ -0,0 +1,142 @@
+package ec2
+
+import (
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+
+	"github.com/crossplane/provider-aws/apis/ec2/v1alpha1"
+	awsclient "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+const (
+	errCIDRAssociationNotFound = "CIDRBlock.NotFound"
+)
+
+// VPCCIDRBlockClient is the external client used for VPC CIDR Block Custom Resource
+type VPCCIDRBlockClient interface {
+	DescribeVpcsRequest(*ec2.DescribeVpcsInput) ec2.DescribeVpcsRequest
+	AssociateVpcCidrBlockRequest(*ec2.AssociateVpcCidrBlockInput) ec2.AssociateVpcCidrBlockRequest
+	DisassociateVpcCidrBlockRequest(*ec2.DisassociateVpcCidrBlockInput) ec2.DisassociateVpcCidrBlockRequest
+}
+
+// NewVPCCIDRBlockClient returns a new client using AWS credentials as JSON encoded data.
+func NewVPCCIDRBlockClient(cfg aws.Config) VPCCIDRBlockClient {
+	return ec2.New(cfg)
+}
+
+// CIDRNotFoundError will be raised when there is no Association
+type CIDRNotFoundError struct{}
+
+// Error satisfies the Error interface for CIDRNotFoundError.
+func (r *CIDRNotFoundError) Error() string {
+	return errCIDRAssociationNotFound
+}
+
+// IsCIDRNotFound returns true if the error code indicates that the CIDR Block Associtation was not found
+func IsCIDRNotFound(err error) bool {
+	_, ok := err.(*CIDRNotFoundError)
+	return ok
+}
+
+// IsVpcCidrBlockUpToDate returns true if there is no update-able difference between desired
+// and observed state of the resource.
+func IsVpcCidrBlockUpToDate(associationID string, spec v1alpha1.VPCCIDRBlockParameters, vpc ec2.Vpc) (bool, error) {
+	IPv4, IPv6 := FindCIDRAssociation(associationID, vpc)
+
+	if IPv4 != nil {
+		return *spec.CIDRBlock == *IPv4.CidrBlock, nil
+	}
+
+	if IPv6 != nil {
+		return aws.StringValue(spec.IPv6CIDRBlock) == aws.StringValue(IPv6.Ipv6CidrBlock) &&
+			aws.StringValue(spec.IPv6Pool) == aws.StringValue(IPv6.Ipv6Pool) &&
+			aws.StringValue(spec.IPv6CIDRBlockNetworkBorderGroup) == aws.StringValue(IPv6.NetworkBorderGroup), nil
+	}
+	return false, &CIDRNotFoundError{}
+}
+
+// IsVpcCidrDeleting returns true if the CIDR Block is already disassociated or disassociating
+func IsVpcCidrDeleting(observation v1alpha1.VPCCIDRBlockObservation) bool {
+	if observation.CIDRBlockState == nil && observation.IPv6CIDRBlockState == nil {
```

Comment:
Can we use a single `switch` for all three conditions in this function for better readability?